File: apps/web/features/shelves/api/schemas.ts
import type { BookshelfItem } from '../../../components/book/bookshelf';
import type { SmartShelfCondition, SmartShelfRules } from '../smart-shelf-rules';
import type { ShelfKind, ShelfView } from '../model/types';
import type { MediaKind } from '../../../types/work';

type JsonObject = Record<string, unknown>;

function isObject(value: unknown): value is JsonObject {
  return typeof value === 'object' && value !== null && !Array.isArray(value);
}

function requiredString(value: unknown, field: string): string {
  if (typeof value !== 'string') throw new Error(`Invalid shelf field: ${field}`);
  return value;
}

function optionalString(value: unknown): string | undefined {
  return typeof value === 'string' ? value : undefined;
}

function optionalNumber(value: unknown): number | undefined {
  return typeof value === 'number' && Number.isFinite(value) ? value : undefined;
}

function stringList(value: unknown): string[] | undefined {
  if (value === undefined) return undefined;
  if (!Array.isArray(value) || value.some((item) => typeof item !== 'string')) {
    throw new Error('Invalid shelf string list');
  }
  return value;
}

function shelfKind(value: unknown): ShelfKind {
  if (value === 'STATIC' || value === 'SMART' || value === 'COLLECTION') return value;
  if (value === undefined) return 'STATIC';
  throw new Error('Invalid shelf kind');
}

function mediaKinds(value: unknown): MediaKind[] {
  if (!Array.isArray(value)) throw new Error('Invalid shelf book media kinds');
  return value.map((item) => {
    if (item === 'EBOOK' || item === 'COMIC' || item === 'AUDIOBOOK') return item;
    throw new Error('Invalid shelf book media kind');
  });
}

function parseBook(value: unknown): BookshelfItem {
  if (!isObject(value)) throw new Error('Invalid shelf book');
  return {
    id: requiredString(value.id, 'book.id'),
    title: requiredString(value.title, 'book.title'),
    author: typeof value.author === 'string' ? value.author : '',
    coverUrl: optionalString(value.coverUrl),
    availableMediaKinds: mediaKinds(value.availableMediaKinds),
    gradient: optionalString(value.gradient),
    coverStatus: optionalString(value.coverStatus)
  };
}

function parseCondition(value: unknown): SmartShelfCondition {
  if (!isObject(value)) throw new Error('Invalid smart shelf condition');
  const rawValue = value.value;
  if (
    rawValue !== undefined
    && typeof rawValue !== 'string'
    && (!Array.isArray(rawValue) || rawValue.some((item) => typeof item !== 'string'))
  ) {
    throw new Error('Invalid smart shelf condition value');
  }
  return {
    field: requiredString(value.field, 'rules.conditions.field'),
    operator: requiredString(value.operator, 'rules.conditions.operator'),
    value: rawValue
  };
}

function parseRules(value: unknown): SmartShelfRules | undefined {
  if (value === undefined || value === null) return undefined;
  if (!isObject(value)) throw new Error('Invalid smart shelf rules');
  const conditions = value.conditions;
  if (conditions !== undefined && !Array.isArray(conditions)) {
    throw new Error('Invalid smart shelf conditions');
  }
  const combinator = value.combinator;
  if (combinator !== undefined && combinator !== 'ALL' && combinator !== 'ANY') {
    throw new Error('Invalid smart shelf combinator');
  }
  return {
    search: optionalString(value.search),
    statuses: stringList(value.statuses),
    mediaKinds: stringList(value.mediaKinds),
    tags: stringList(value.tags),
    authors: stringList(value.authors),
    publishers: stringList(value.publishers),
    combinator,
    conditions: conditions?.map(parseCondition)
  };
}

export function parseShelfView(value: unknown): ShelfView {
  if (!isObject(value)) throw new Error('Invalid shelf response');
  const description = value.description;
  if (description !== undefined && description !== null && typeof description !== 'string') {
    throw new Error('Invalid shelf description');
  }
  if (value.books !== undefined && !Array.isArray(value.books)) {
    throw new Error('Invalid shelf books');
  }
  if (value.shelves !== undefined && !Array.isArray(value.shelves)) {
    throw new Error('Invalid shelf members');
  }
  return {
    id: requiredString(value.id, 'id'),
    name: requiredString(value.name, 'name'),
    description: description ?? null,
    kind: shelfKind(value.kind),
    pinned: typeof value.pinned === 'boolean' ? value.pinned : undefined,
    createdAt: optionalString(value.createdAt) ?? '1970-01-01T00:00:00Z',
    updatedAt: optionalString(value.updatedAt) ?? '1970-01-01T00:00:00Z',
    bookCount: optionalNumber(value.bookCount),
    bookIds: stringList(value.bookIds),
    books: value.books?.map(parseBook),
    collectionIds: stringList(value.collectionIds),
    shelfCount: optionalNumber(value.shelfCount),
    memberShelfIds: stringList(value.memberShelfIds),
    shelves: value.shelves?.map(parseShelfView),
    page: optionalNumber(value.page),
    pageSize: optionalNumber(value.pageSize),
    total: optionalNumber(value.total),
    totalPages: optionalNumber(value.totalPages),
    rules: parseRules(value.rules),
    rulesStatus: value.rulesStatus === 'UNSUPPORTED' ? 'UNSUPPORTED' : 'VALID',
    unsupportedRuleFields: stringList(value.unsupportedRuleFields) ?? []
  };
}

export function parseShelfListPayload(value: unknown): { shelves: ShelfView[] } {
  if (!isObject(value) || !Array.isArray(value.shelves)) {
    throw new Error('Invalid shelf list response');
  }
  return { shelves: value.shelves.map(parseShelfView) };
}

export function parseShelfPayload(value: unknown): { shelf: ShelfView } {
  if (!isObject(value)) throw new Error('Invalid shelf response');
  return { shelf: parseShelfView(value.shelf) };
}

export function parseDeletedShelfPayload(value: unknown): { deleted: boolean; id: string } {
  if (!isObject(value) || typeof value.deleted !== 'boolean') {
    throw new Error('Invalid deleted shelf response');
  }
  return {
    deleted: value.deleted,
    id: requiredString(value.id, 'id')
  };
}

export function parseShelfEnvelope(
  value: unknown
): { ok: boolean; data?: unknown; error?: { message?: string; code?: string } } {
  if (!isObject(value) || typeof value.ok !== 'boolean') {
    throw new Error('Invalid shelf API envelope');
  }
  const error = isObject(value.error)
    ? {
        message: optionalString(value.error.message),
        code: optionalString(value.error.code)
      }
    : undefined;
  return { ok: value.ok, data: value.data, error };
}

File: apps/api-python/app/modules/shelf/presentation/schemas.py
from __future__ import annotations

from datetime import datetime
from typing import Literal

from pydantic import Field

from app.contracts.http import HttpContractModel, SuccessEnvelope
from app.modules.reader.public import MediaKind

FilterValue = str | int | float | bool | list[str] | None


class ShelfCondition(HttpContractModel):
    field: str
    operator: str
    value: FilterValue = None


class ShelfRules(HttpContractModel):
    search: str | None = None
    statuses: list[str] | None = None
    media_kinds: list[str] | None = Field(default=None, alias="mediaKinds")
    tags: list[str] | None = None
    authors: list[str] | None = None
    publishers: list[str] | None = None
    combinator: Literal["ALL", "ANY"] | None = None
    conditions: list[ShelfCondition] | None = None
    included_work_ids: list[str] | None = Field(default=None, alias="includedWorkIds")


class ShelfWriteRequest(HttpContractModel):
    name: str | None = None
    description: str | None = None
    kind: Literal["STATIC", "SMART", "COLLECTION"] | None = None
    rules: ShelfRules | None = None
    pinned: bool | None = None
    book_ids: list[str] | None = Field(default=None, alias="bookIds")
    work_ids: list[str] | None = Field(default=None, alias="workIds")
    collection_ids: list[str] | None = Field(default=None, alias="collectionIds")
    member_shelf_ids: list[str] | None = Field(
        default=None,
        alias="memberShelfIds",
    )


class ShelfBook(HttpContractModel):
    id: str
    title: str
    author: str
    cover_url: str = Field(alias="coverUrl")
    available_media_kinds: list[MediaKind] = Field(alias="availableMediaKinds")


class ShelfMemberView(HttpContractModel):
    id: str
    name: str
    description: str | None
    kind: Literal["STATIC", "SMART"]
    pinned: bool
    book_count: int = Field(alias="bookCount")
    books: list[ShelfBook]
    collection_ids: list[str] = Field(alias="collectionIds")
    created_at: datetime = Field(alias="createdAt")
    updated_at: datetime = Field(alias="updatedAt")


class ShelfView(HttpContractModel):
    id: str
    owner_user_id: str | None = Field(default=None, alias="ownerUserId")
    name: str
    description: str | None
    kind: Literal["STATIC", "SMART", "COLLECTION"]
    rules_json: str = Field(alias="rulesJson")
    pinned: bool
    created_at: datetime = Field(alias="createdAt")
    updated_at: datetime = Field(alias="updatedAt")
    rules: ShelfRules
    rules_status: Literal["VALID", "UNSUPPORTED"] = Field(alias="rulesStatus")
    unsupported_rule_fields: list[str] = Field(alias="unsupportedRuleFields")
    book_count: int | None = Field(
        default=None,
        alias="bookCount",
        exclude_if=lambda value: value is None,
    )
    books: list[ShelfBook] | None = Field(
        default=None,
        exclude_if=lambda value: value is None,
    )
    collection_ids: list[str] | None = Field(
        default=None,
        alias="collectionIds",
        exclude_if=lambda value: value is None,
    )
    shelf_count: int | None = Field(
        default=None,
        alias="shelfCount",
        exclude_if=lambda value: value is None,
    )
    shelves: list[ShelfMemberView] | None = Field(
        default=None,
        exclude_if=lambda value: value is None,
    )
    member_shelf_ids: list[str] | None = Field(
        default=None,
        alias="memberShelfIds",
        exclude_if=lambda value: value is None,
    )
    page: int | None = None
    page_size: int | None = Field(default=None, alias="pageSize")
    total: int | None = None
    total_pages: int | None = Field(default=None, alias="totalPages")
    book_ids: list[str] | None = Field(
        default=None,
        alias="bookIds",
        exclude_if=lambda value: value is None,
    )


class ShelvesPayload(HttpContractModel):
    shelves: list[ShelfView]


class ShelfPayload(HttpContractModel):
    shelf: ShelfView


class DeletedShelfPayload(HttpContractModel):
    deleted: bool
    id: str


ShelvesResponse = SuccessEnvelope[ShelvesPayload]
ShelfResponse = SuccessEnvelope[ShelfPayload]
DeletedShelfResponse = SuccessEnvelope[DeletedShelfPayload]
